### PR TITLE
Add the EC2 Istance ID to AWS Xray traces

### PR DESCRIPTION
**Pull request by joehoyle**
Per the docs at https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-aws

### CODE REVIEW

File: inc/namespace.php [hunk 1]
```diff
@@ -456,3 +459,92 @@ function on_request_stats( TransferStats $stats ) {
...
+ *    privateIp: string,
+ *    ramdiskId: string,
+ *    region: string,
+ *    version: string,
+ * )
+ */
+function get_ec2_instance_metadata() : array {
+	$has_cache = false;
+	$cache_key = 'altis.ec2_instance_metadata';
+	$cached_data = apcu_fetch( $cache_key, $has_cache );
```
Comment (roborourke) [inline inc/namespace.php]: For my own understanding why do we need to use `apcu` for this?

Also it looks like the `apcu` package was only just added - should we add a `function_exists()` check here?

File: inc/namespace.php [hunk 2]
```diff
@@ -456,3 +459,92 @@ function on_request_stats( TransferStats $stats ) {
...
+		return [];
+	}
+
+	if ( $request->getStatusCode() !== '200' ) {
+		trigger_error( sprintf( 'Unable to get instance metadata. Returned response code: %s', $request->getStatusCode() ), E_USER_NOTICE );
+		apcu_store( $cache_key, [] );
+		return [];
+	}
+
+	$metadata = json_decode( $request->getBody(), true );
```
Comment (roborourke) [inline inc/namespace.php]: Looks like you can just do `$request->json();` here

File: inc/namespace.php [hunk 3]
```diff
@@ -456,3 +459,92 @@ function on_request_stats( TransferStats $stats ) {
...
+	if ( ! empty( $trace['parent_id'] ) ) {
+		return $trace;
+	}
+
+	$metadata = get_ec2_instance_metadata();
+	if ( ! $metadata ) {
+		return $trace;
+	}
+
+	$trace['aws']['ec2']['availability_zone'] = $metadata['availabilityZone'];
```
Comment (roborourke) [inline inc/namespace.php]: It's a bit confusing that we can get both ec2 and ecs instance metadata, I thought they were somewhat exclusive from an infrastructure standpoint. Can you add a few comments to note why there's ec2 data when using ecs / where it comes from? Will there be ec2 metadata when using ecs?

File: inc/namespace.php (hunk 1 above)
Comment (rmccue) [reply inc/namespace.php]: This is so it's specifically a per-machine cache I think, whereas the object cache is per-stack.

We should already have had apcu installed I think? We're using it on humanmade-production-v2, so it's been around for a while afaik.

Comment (joehoyle) [reply inc/namespace.php]: Yeah apcu has been there for some time, we just didn't have it documented. It is indeed to cache per-machine. I'll add an inline comment.

File: inc/namespace.php (hunk 3 above)
Comment (joehoyle) [reply inc/namespace.php]: No, ecs runs on ec2, so in the case of running on ecs, we want to send both the underlaying instance data, and the ecs container id.

Comment (roborourke) [reply inc/namespace.php]: Ah cool, makes sense of course. I thought it was maybe a bit more abstracted than that 👍

**Comment by joehoyle** [comment]
Blocked by https://github.com/humanmade/aws-xray/pull/55

File: inc/namespace.php (hunk 2 above)
Comment (joehoyle) [reply inc/namespace.php]: Hmm atleast in the version of Guzzle we have, it looks like that isn't in `MessageInterface`

Comment (roborourke) [reply inc/namespace.php]: Sad times 😢 As you were...

**Comment by joehoyle** [comment]
Ok unblocked, ready for final review.

File: inc/namespace.php [hunk 4]
```diff
@@ -456,3 +459,97 @@ function on_request_stats( TransferStats $stats ) {
...
+	// not across all servers (wp_cache_*).
+	$cached_data = apcu_fetch( $cache_key, $has_cache );
+	if ( $has_cache ) {
+		return $cached_data;
+	}
+
+	$client = new Client();
+
+	try {
+		$request = $client->request( 'GET', 'http://169.254.169.254/latest/dynamic/instance-identity/document', [
```
Comment (roborourke) [inline inc/namespace.php]: Any value in documenting where this IP comes from? Assuming it's a standard AWS thing.

Comment (rmccue) [reply inc/namespace.php]: Yep, it's the standard link-local IP: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html

Comment (joehoyle) [reply inc/namespace.php]: IMO this IP is common enough I think.

**Pull request merged**